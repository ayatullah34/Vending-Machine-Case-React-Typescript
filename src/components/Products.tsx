import React, { useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { BiTimer } from 'react-icons/bi';
import { useDispatch, useSelector } from 'react-redux';
import { toast } from 'react-toastify';
import { AnyAction, Dispatch } from 'redux';
import { Product } from '../interfaces/IProduct';
import { RootState } from '../interfaces/IRootState';
import { setCoinTotal, setSelectedProducts } from '../redux/machineSlice';
import { formatTime } from '../constant/commonFunc';
import { TimerProps } from '../interfaces/ITimer';
import { RESET_TIME_IN_MS, products } from '../data/products';

function Products({ timeLeft = 0, startTimer = () => { }, resetTimer = () => { } }: TimerProps) {
    const dispatch: Dispatch<AnyAction> = useDispatch();
    const { t } = useTranslation();
    const [productItems, setProductItems] = useState(products);
    const resetProductItem = useSelector((state: RootState) => state.machine.resetProductItem)
    const selectedProducts = useSelector((state: RootState) => state.machine.selectedProducts)
    const [isAlmostZero, setIsAlmostZero] = useState<boolean>(false);

    const handleProductClick = (item: Product) => {
        const selectedProductIndex = selectedProducts.findIndex(
            (product) => product.name === item.name
        );

        if (selectedProductIndex > -1) {
            const updatedSelectedProducts = [...selectedProducts];
            updatedSelectedProducts.splice(selectedProductIndex, 1);
            dispatch(setSelectedProducts(updatedSelectedProducts));
            setProductItems((prev) =>
                prev.map((product) =>
                    product.name === item.name ? { ...product, disabled: false } : product
                )
            );
        } else {
            const updatedSelectedProducts = [...selectedProducts, item];
            dispatch(setSelectedProducts(updatedSelectedProducts));
            setProductItems((prev) =>
                prev.map((product) =>
                    product.name === item.name ? { ...product, disabled: true } : product
                )
            );
        }
    };


    useEffect(() => {
        if (Object.keys(selectedProducts || []).length > 0) {
            if (startTimer) {
                startTimer();
            }
            const disableActions = () => {
                dispatch(setSelectedProducts([]));
                dispatch(setCoinTotal(0));
                setProductItems((prevItems) =>
                    prevItems.map((item) => ({
                        ...item,
                        disabled: false,
                    }))
                );
                toast.info(t("user_interaction_end"), {
                    autoClose: 2000,
                });
            };

            const interactionInterval = setInterval(() => {
                disableActions();
            }, RESET_TIME_IN_MS);

            return () => {
                clearInterval(interactionInterval);
            };
        } else {
            resetTimer();
        }
    }, [selectedProducts]);

    useEffect(() => {
        if (timeLeft) {
            if (timeLeft <= 5000) {
                setIsAlmostZero(true);
            }
            else {
                setIsAlmostZero(false);
            }
        }
    }, [timeLeft]);

    // Tempature products set to min 6Â°C
    useEffect(() => {
        const intervalId = setInterval(() => {
            setProductItems((prev) =>
                prev.map((product) =>
                    product.temperature >= 11
                        ? { ...product, temperature: product.temperature - 5 }
                        : product
                )
            );
        }, 5 * 60 * 1000);

        return () => clearInterval(intervalId);
    }, []);

    useEffect(() => {
        setProductItems(products);
    }, [resetProductItem])

    return (
        <div className="vending-machine-container">
            <h1 className={isAlmostZero ? "red-label" : ""}><BiTimer /> {timeLeft && formatTime(timeLeft)}</h1>
            <div className="items">
                {productItems.map((item) => (
                    <div
                        data-i18n-refund={t("refund")}
                        className={`items_box ${item.disabled ? "disabled" : ""}`}
                        key={item.name}
                        onClick={() => handleProductClick(item)}
                    >
                        <img
                            src={`${process.env.PUBLIC_URL}/assets/products/${item.imgSrc}`}
                            alt={item.name}
                        />
                        <div>{t(item.name)}</div>
                        <div>
                            {item.currency}
                            {item.price}
                        </div>
                    </div>
                ))}
            </div>
        </div>)
}

export default Products;