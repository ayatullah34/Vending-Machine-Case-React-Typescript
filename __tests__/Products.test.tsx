import "@testing-library/jest-dom/extend-expect";
import React from "react";
import { render, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import Products from '../src/components/Products';

const mockStore = configureMockStore([]);
const resetTimer = jest.fn();

describe('Products', () => {
    let store;
    let component;

    beforeEach(() => {
        store = mockStore({
            machine: {
                resetProductItem: false,
                selectedProducts: []
            }
        });

        component = render(
            <Provider store={store}>
                <Products resetTimer={resetTimer} />
            </Provider>
        );
    });

    it('renders correctly', () => {
        expect(component).toMatchSnapshot();
    });

    it('handles product click correctly', () => {
        const productItem = component.getByText('taco');
        fireEvent.click(productItem);
        expect(store.getActions()).toEqual([
            {
                type: 'machine/setSelectedProducts',
                payload: [{ name: "taco", price: 60, currency: "$", imgSrc: "taco.svg", disabled: false, temperature: 6}]
            }
        ]);
    });
});
