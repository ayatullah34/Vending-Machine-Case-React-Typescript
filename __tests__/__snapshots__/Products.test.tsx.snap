// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Products renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="vending-machine-container"
      >
        <h1
          class=""
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m20.145 8.27 1.563-1.563-1.414-1.414L18.586 7c-1.05-.63-2.274-1-3.586-1-3.859 0-7 3.14-7 7s3.141 7 7 7 7-3.14 7-7a6.966 6.966 0 0 0-1.855-4.73zM15 18c-2.757 0-5-2.243-5-5s2.243-5 5-5 5 2.243 5 5-2.243 5-5 5z"
            />
            <path
              d="M14 10h2v4h-2zm-1-7h4v2h-4zM3 8h4v2H3zm0 8h4v2H3zm-1-4h3.99v2H2z"
            />
          </svg>
          0
        </h1>
        <div
          class="items"
        >
          <div
            class="items_box "
            data-i18n-refund="refund"
          >
            <img
              alt="cola"
              src="undefined/assets/products/cola.svg"
            />
            <div>
              cola
            </div>
            <div>
              $
              25
            </div>
          </div>
          <div
            class="items_box "
            data-i18n-refund="refund"
          >
            <img
              alt="chips"
              src="undefined/assets/products/chips.svg"
            />
            <div>
              chips
            </div>
            <div>
              $
              20
            </div>
          </div>
          <div
            class="items_box "
            data-i18n-refund="refund"
          >
            <img
              alt="candy"
              src="undefined/assets/products/candy.svg"
            />
            <div>
              candy
            </div>
            <div>
              $
              15
            </div>
          </div>
          <div
            class="items_box "
            data-i18n-refund="refund"
          >
            <img
              alt="popcorn"
              src="undefined/assets/products/popcorn.svg"
            />
            <div>
              popcorn
            </div>
            <div>
              $
              30
            </div>
          </div>
          <div
            class="items_box "
            data-i18n-refund="refund"
          >
            <img
              alt="water"
              src="undefined/assets/products/water.svg"
            />
            <div>
              water
            </div>
            <div>
              $
              5
            </div>
          </div>
          <div
            class="items_box "
            data-i18n-refund="refund"
          >
            <img
              alt="chocolate"
              src="undefined/assets/products/chocolate.svg"
            />
            <div>
              chocolate
            </div>
            <div>
              $
              10
            </div>
          </div>
          <div
            class="items_box "
            data-i18n-refund="refund"
          >
            <img
              alt="bread"
              src="undefined/assets/products/bread.svg"
            />
            <div>
              bread
            </div>
            <div>
              $
              5
            </div>
          </div>
          <div
            class="items_box "
            data-i18n-refund="refund"
          >
            <img
              alt="taco"
              src="undefined/assets/products/taco.svg"
            />
            <div>
              taco
            </div>
            <div>
              $
              60
            </div>
          </div>
          <div
            class="items_box "
            data-i18n-refund="refund"
          >
            <img
              alt="hotdog"
              src="undefined/assets/products/hotdog.svg"
            />
            <div>
              hotdog
            </div>
            <div>
              $
              40
            </div>
          </div>
          <div
            class="items_box "
            data-i18n-refund="refund"
          >
            <img
              alt="coffee"
              src="undefined/assets/products/coffee.svg"
            />
            <div>
              coffee
            </div>
            <div>
              $
              50
            </div>
          </div>
        </div>
        <div
          class="leds"
        >
          <div
            class="tooltip-container"
          >
            <div
              class="leds_green"
            />
          </div>
          <div
            class="tooltip-container"
          >
            <div
              class="leds_blue  "
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="vending-machine-container"
    >
      <h1
        class=""
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m20.145 8.27 1.563-1.563-1.414-1.414L18.586 7c-1.05-.63-2.274-1-3.586-1-3.859 0-7 3.14-7 7s3.141 7 7 7 7-3.14 7-7a6.966 6.966 0 0 0-1.855-4.73zM15 18c-2.757 0-5-2.243-5-5s2.243-5 5-5 5 2.243 5 5-2.243 5-5 5z"
          />
          <path
            d="M14 10h2v4h-2zm-1-7h4v2h-4zM3 8h4v2H3zm0 8h4v2H3zm-1-4h3.99v2H2z"
          />
        </svg>
        0
      </h1>
      <div
        class="items"
      >
        <div
          class="items_box "
          data-i18n-refund="refund"
        >
          <img
            alt="cola"
            src="undefined/assets/products/cola.svg"
          />
          <div>
            cola
          </div>
          <div>
            $
            25
          </div>
        </div>
        <div
          class="items_box "
          data-i18n-refund="refund"
        >
          <img
            alt="chips"
            src="undefined/assets/products/chips.svg"
          />
          <div>
            chips
          </div>
          <div>
            $
            20
          </div>
        </div>
        <div
          class="items_box "
          data-i18n-refund="refund"
        >
          <img
            alt="candy"
            src="undefined/assets/products/candy.svg"
          />
          <div>
            candy
          </div>
          <div>
            $
            15
          </div>
        </div>
        <div
          class="items_box "
          data-i18n-refund="refund"
        >
          <img
            alt="popcorn"
            src="undefined/assets/products/popcorn.svg"
          />
          <div>
            popcorn
          </div>
          <div>
            $
            30
          </div>
        </div>
        <div
          class="items_box "
          data-i18n-refund="refund"
        >
          <img
            alt="water"
            src="undefined/assets/products/water.svg"
          />
          <div>
            water
          </div>
          <div>
            $
            5
          </div>
        </div>
        <div
          class="items_box "
          data-i18n-refund="refund"
        >
          <img
            alt="chocolate"
            src="undefined/assets/products/chocolate.svg"
          />
          <div>
            chocolate
          </div>
          <div>
            $
            10
          </div>
        </div>
        <div
          class="items_box "
          data-i18n-refund="refund"
        >
          <img
            alt="bread"
            src="undefined/assets/products/bread.svg"
          />
          <div>
            bread
          </div>
          <div>
            $
            5
          </div>
        </div>
        <div
          class="items_box "
          data-i18n-refund="refund"
        >
          <img
            alt="taco"
            src="undefined/assets/products/taco.svg"
          />
          <div>
            taco
          </div>
          <div>
            $
            60
          </div>
        </div>
        <div
          class="items_box "
          data-i18n-refund="refund"
        >
          <img
            alt="hotdog"
            src="undefined/assets/products/hotdog.svg"
          />
          <div>
            hotdog
          </div>
          <div>
            $
            40
          </div>
        </div>
        <div
          class="items_box "
          data-i18n-refund="refund"
        >
          <img
            alt="coffee"
            src="undefined/assets/products/coffee.svg"
          />
          <div>
            coffee
          </div>
          <div>
            $
            50
          </div>
        </div>
      </div>
      <div
        class="leds"
      >
        <div
          class="tooltip-container"
        >
          <div
            class="leds_green"
          />
        </div>
        <div
          class="tooltip-container"
        >
          <div
            class="leds_blue  "
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
